### PR TITLE
MWPW-136891: Introduce Switch Modal feature

### PR DESCRIPTION
**Pull request by 3ch023**
### Description
Switch Modal is also known as Upgrade Flow.
For each commerce CTA check if a user is logged in and eligible for an upgrade, if so - Commerce CTA will open a modal with Iframe to the [Switch Plan](https://wiki.corp.adobe.com/pages/viewpage.action?spaceKey=IdentityACCM&title=Manage+Plan+Integration).

To enable the feature on the page GWP needs to add a `merch-offers` block containing the Upgrade Offer.
The block's position doesn't matter, but we recommend placing it at the top or bottom of the page for simplicity.
image

This implementation doesn't have a pre-spinner implemented, this will be done in [MWPW-142250](https://jira.corp.adobe.com/browse/MWPW-142250)

### How to test

1.  Install ModHeader to avoid CORS issues with WCS and Switch Plan
2. Contact me for ModHeader profiles and import them in your plugin
3. For IMS lib not working on feature branches install 'CORS Unblock'
4. Login with a test user (contact me for creds)
5. Enable 'CORS Unblock' and ModHeader 'wcs' profile
6. Load the [Test Page](https://ims--milo--adobecom.hlx.page/drafts/mariia/upgradeFlow/upgradeflow)
7. For IMS lib not working on feature branches

Resolves: [MWPW-136891](https://jira.corp.adobe.com/browse/MWPW-136891)

### Consumer Details
The feature is disabled by default.
CC: https://main--cc--adobecom.hlx.page/products/special-offers?milolibs=ims 
DC: https://main--dc--adobecom.hlx.page/drafts/mariia/document?milolibs=ims
Homepage: - no merch yet

image

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/mariia/upgradeFlow/upgradeflow?martech=off&georouting=off
- After: https://ims--milo--adobecom.hlx.page/drafts/mariia/upgradeFlow/upgradeflow?martech=off&georouting=off

### CODE REVIEW

File: libs/blocks/merch/merch.css [hunk 1]
```diff
@@ -29,3 +29,13 @@ a[is='checkout-link'].loading-entitlements > span {
 .price-tax-inclusivity:not(.disabled)::before {
   content: "\00a0";
 }
+
+.upgrade-flow-iframe {
```
Comment (yesil) [inline libs/blocks/merch/merch.css:33]: is this supposed to be responsive?

File: libs/blocks/merch/upgrade.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
+import { createTag, getConfig } from '../../utils/utils.js';
+import { replaceKey } from '../../features/placeholders.js';
+
+const SOURCE_PF = ['PHOTOSHOP', 'ILLUSTRATOR', 'ACROBAT', 'AUDITION', 'RUSH', 'INDESIGN', '3DI', 'XD', 'PREMIERE', 'AFTEREFFECTS', 'DREAMWEAVER', 'CC_EXPRESS', 'INCOPY', '3D_TEXTURING', 'PHOTOGRAPHY', 'EDGE_ANIMATE', 'PHOTOSHOP_LIGHTROOM', 'ILLUSTRATOR_STOCK_BUNDLE'];
```
Comment (yesil) [inline libs/blocks/merch/upgrade.js:4]: 👍🏻

File: libs/blocks/merch/upgrade.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  RETURN_BACK: 'RETURN_BACK',
+  OrderComplete: 'OrderComplete',
+  Error: 'Error',
+  Close: 'Close',
+};
+const isProductFamily = (offer, pfs) => {
+  const productFamily = offer?.offer?.product_arrangement?.family;
+  return productFamily && pfs.includes(productFamily);
+};
+let shouldRefetchEntitlements = false;
```
Comment (yesil) [inline libs/blocks/merch/upgrade.js:19]: nit: consistent spacing, empty line before line 19?

File: libs/blocks/merch/upgrade.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      window.open(data.externalUrl, data.target);
+      break;
+    case MANAGE_PLAN_MSG_SUBTYPE.SWITCH:
+      if (!data?.externalUrl || !data.target) return;
+      window.open(data.externalUrl, data.target);
+      break;
+    case MANAGE_PLAN_MSG_SUBTYPE.RETURN_BACK:
+      if (!data?.externalUrl || !data.target) return;
+      if (data.returnUrl) {
+        window.sessionStorage.setItem('upgradeModalReturnUrl', data.returnUrl);
```
Comment (yesil) [inline libs/blocks/merch/upgrade.js:70]: ℹ️

**Comment by yesil** [comment]
@3ch023 fyi on mobile, it doesn't fit in the modal.

File: libs/blocks/merch/upgrade.js (hunk 2 above)
Comment (3ch023) [reply libs/blocks/merch/upgrade.js:4]: :)

File: libs/blocks/merch/merch.js [hunk 5]
```diff
@@ -37,16 +37,33 @@ export async function loadEntitlements() {
...
   });
   return loadEntitlements.promise;
 }
 
 async function getCheckoutAction(offers) {
-  const [{ arrangement_codes: aCodes } = {}, entitlementsMappings] = await loadEntitlements();
-  if (aCodes === undefined) return undefined;
+  let upgradeAction;
+  const [entitlements, entitlementsMappings] = await loadEntitlements();
+  const aCodes = entitlements?.map((offer) => offer.offer.product_arrangement_code);
```
Comment (npeltier) [inline libs/blocks/merch/merch.js:48]: is `product_arrangement_code` always supposed to be here?

File: libs/blocks/merch/merch.js [hunk 6]
```diff
@@ -37,16 +37,33 @@ export async function loadEntitlements() {
...
   const [{ productArrangementCode }] = offers;
-  if (aCodes[productArrangementCode] === true) {
+  const [{ productArrangement }] = offers;
```
Comment (npeltier) [inline libs/blocks/merch/merch.js:50]: `const [{productArrangementCode, productArrangement}] = offers;` ?

File: libs/blocks/merch/merch.js [hunk 7]
```diff
@@ -37,16 +37,33 @@ export async function loadEntitlements() {
...
+  const upgradeOffer = await document.querySelector('.merch-offers.upgrade [data-wcs-osi]')?.onceSettled()
+    .catch((e) => {
+      window.lana.log('Failed to resolve an upgrade offer:', e);
+      return undefined;
+    });
+  if (upgradeOffer && Array.isArray(entitlements) && ctaProductFamily) {
+    const { default: handleUpgradeOffer } = await import('./upgrade.js');
+    upgradeAction = handleUpgradeOffer(ctaProductFamily, upgradeOffer, entitlements);
+  }
+  if (upgradeAction) return upgradeAction;
```
Comment (npeltier) [inline libs/blocks/merch/merch.js:61]: whole reason why this test can be false and this strange 57-61 block logic is here, is because handleUpgradeOffer could get from buildUrl that both upgrade offer & origin offer have an id. Can't we make this cleaner / simpler by extracting the source & target offer extraction from entitlement logic and line 57 testing if there are any, then line 59 we know we have all that we need and we can return the url.
In other words, could handleUpgradeOffer just keep the offer extraction logic, and return target offer (or undefined),
and could we return buildURL l59? 

wdyt?

File: libs/blocks/merch/merch.js (hunk 6 above)
Comment (mirafedas) [inline libs/blocks/merch/merch.js:50]: As we discussed in a huddle:
```
async function getCheckoutAction(offers) {
  const [entitlements, entitlementsMappings] = await loadEntitlements();
  if (offers?.length) {
  const [{ productArrangementCode, productArrangement }] = offers || [];
  const { productFamily } = productArrangement?.productFamily || {};
  const upgradeOffer = await document.querySelector('.merch-offers.upgrade [data-wcs-osi]')?.onceSettled()
    .catch((e) => {
      window.lana.log('Failed to resolve an upgrade offer:', e);
      return undefined;
    });
  if (upgradeOffer && entitlements?.length && productFamily) {
    const { default: handleUpgradeOffer } = await import('./upgrade.js');
    const upgradeAction = handleUpgradeOffer(productFamily, upgradeOffer, entitlements);
   if (upgradeAction) return upgradeAction;
  }
 }
 
  const downloadFlow = false; // temporary disable
  if (downloadFlow === false) return undefined;
  const aCodes = entitlements?.map((offer) => offer.offer.product_arrangement_code);
  if (aCodes?.includes(productArrangementCode)) {
    const mapping = entitlementsMappings.data
      ?.find(({ [TITLE_PRODUCT_ARRANGEMENT_CODE]: code }) => code === productArrangementCode);
    if (!mapping) return undefined;
    const config = getConfig();
    const { locale: { ietf } } = config;
    const text = await replaceKey(mapping.CTA, config);
    const url = mapping[ietf] || localizeLink(mapping.Target);
    return { text, url };
  }
  return undefined;
}
```

**Review by mirafedas — CHANGES_REQUESTED** [review]

File: libs/blocks/merch/merch.js (hunk 5 above)
Comment (3ch023) [reply libs/blocks/merch/merch.js:48]: no, this is download flow (different feature), will be fixed in a separate ticket
@yesil should have a jira number

File: libs/blocks/merch/merch.css (hunk 1 above)
Comment (3ch023) [reply libs/blocks/merch/merch.css:33]: will fix styles

File: libs/blocks/merch/merch.js (hunk 7 above)
Comment (3ch023) [reply libs/blocks/merch/merch.js:61]: sorry, i didn't understand your suggestion.
do you mean to move the code below into handleUpgradeOffer in `upgrade.js`?
```
const upgradeOffer = await document.querySelector('.merch-offers.upgrade [data-wcs-osi]')?.onceSettled()
    .catch((e) => {
      window.lana.log('Failed to resolve an upgrade offer:', e);
      return undefined;
    });
```

>In other words, could handleUpgradeOffer just keep the offer extraction logic, and return target offer (or undefined),

handleUpgradeOffer returns an action object which contains cta text, and a function that renders the modal. In your suggestion if only URL is returned by handleUpgradeOffer, where will the modal be created?

File: libs/blocks/merch/merch.js (hunk 6 above)
Comment (3ch023) [reply libs/blocks/merch/merch.js]: fixed

**Comment by 3ch023** [comment]
> @3ch023 fyi on mobile, it doesn't fit in the modal.

thx, add styles for mobile/tablet same as we have it on dexter

File: libs/blocks/merch/merch.js (hunk 6 above)
Comment (mirafedas) [reply libs/blocks/merch/merch.js]: Sorry, I forgot to remove one redundant checking, instead of:
 ```const [{ productArrangementCode, productArrangement }] = offers || [];```
 should be:
  ```const [{ productArrangementCode, productArrangement }] = offers;```

File: libs/blocks/merch/merch.js (hunk 7 above)
Comment (3ch023) [reply libs/blocks/merch/merch.js]: as discussed offline, I could move the checks for eligibility to merch.js and have them all run before we load upgrade.js, but then part of upgrade logic will be in one file and another part in another file.
which is ok for me, although I lean more towards keeping as much as possible in upgrade.js
would like to know also what others think.

File: libs/blocks/merch/merch.js (hunk 6 above)
Comment (3ch023) [reply libs/blocks/merch/merch.js]: yep, this is what i have atm

**Review by npeltier — APPROVED** [review]
lgtm

**Review by mirafedas — APPROVED** [review]

**Review by yesil — APPROVED** [review]
